contracts/verifying/VerifyingSingletonPaymasterV2.sol

```diff
@@ -0,0 +1,390 @@
+        (
+            address paymasterId,
+            uint48 validUntil,
+            uint48 validAfter,
+            uint32 priceMarkup,
+            bytes calldata signature
+        ) = parsePaymasterAndData(userOp.paymasterAndData);
+
+        bytes32 hash = getHash(
+            userOp,
+            paymasterId,
+            validUntil,
+            validAfter,
+            priceMarkup
+        );
+        uint256 sigLength = signature.length;
+        // we only "require" it here so that the revert reason on invalid signature will be of "VerifyingPaymaster", and not "ECDSA"
+        if (sigLength != 65) revert InvalidPaymasterSignatureLength(sigLength);
+        //don't revert on signature failure: return SIG_VALIDATION_FAILED
+        if (
+            verifyingSigner != hash.toEthSignedMessageHash().recover(signature)
+        ) {
+            // empty context and sigFailed with time range provided
+            return (context, _packValidationData(true, validUntil, validAfter));
+        }
+
+        require(priceMarkup <= 2e6, "Verifying PM:high markup %");
+
+        uint32 dynamicMarkup = maxuint32(priceMarkup, fixedPriceMarkup);
+        require(dynamicMarkup >= 1e6, "Verifying PM:low markup %");
+
+        uint256 effectiveCost = (requiredPreFund * dynamicMarkup) /
+            PRICE_DENOMINATOR;
+
+        if (effectiveCost > paymasterIdBalances[paymasterId])
+            revert InsufficientBalance(
+                effectiveCost,
+                paymasterIdBalances[paymasterId]
+            );
+
+        context = abi.encode(
+            paymasterId,
+            priceMarkup,
+            userOp.maxFeePerGas,
+            userOp.maxPriorityFeePerGas
+        );
+
+        return (context, _packValidationData(false, validUntil, validAfter));
+    }
+
+    function parsePaymasterAndData(
+        bytes calldata paymasterAndData
+    )
+        public
+        pure
+        returns (
+            address paymasterId,
+            uint48 validUntil,
+            uint48 validAfter,
+            uint32 priceMarkup,
+            bytes calldata signature
+        )
+    {
+        (paymasterId, validUntil, validAfter, priceMarkup) = abi.decode(
+            paymasterAndData[VALID_PND_OFFSET:SIGNATURE_OFFSET],
+            (address, uint48, uint48, uint32)
+        );
+
+        signature = paymasterAndData[SIGNATURE_OFFSET:];
+    }
+
+    /**
+     * @dev Executes the paymaster's payment conditions
+     * @param mode tells whether the op succeeded, reverted, or if the op succeeded but cause the postOp to revert
+     * @param context payment conditions signed by the paymaster in `validatePaymasterUserOp`
+     * @param actualGasCost amount to be paid to the entry point in wei
+     */
+    function _postOp(
+        PostOpMode mode,
+        bytes calldata context,
+        uint256 actualGasCost
+    ) internal virtual override {
+        (
+            address paymasterId,
+            uint32 priceMarkup,
```

Comment:
why is price markup uint32? It's more expensive to operate on non uint256 values

Reply:
because it would never exceed 2000000. uint24 also would be enough

Reply:
makes sense, but why do you want to pay extra gas for artificially constraining it to uint32?

Reply:
keeping as it is but will test changes once to double check if there are significant savings

Reply:
I ran the tests. You save about `45` gas in `validatePaymasterUserOp` and `70` gas in `postOp`